rfc-117-container-runtime.md

```diff
@@ -0,0 +1,75 @@
+# RFC 117: Use a container runtime
+
+## Summary
+
+Use a container runtime for hosting GOV.UK applications. This will provide a number of benefits, including:
+* Removing our dependency on a version of Ubuntu (Trusty) which is in extended support (which ends in April 2022)
+* Removing legacy infrastructure code which is difficult to maintain and extend
+* Providing better value for money by using cloud resources more effectively
+* Improving development velocity by reducing the difference between production and development environments
+* Making it easier to adopt higher velocity deployment approaches (such as continuous deployment)
+
+## Problem
+
+GOV.UK currently hosts the majority of its applications and services using virtual machines running Ubuntu 14.04 LTS (Trusty Tahr).
+
+Trusty is out of date, it was released in April 2014 and standard support ended in April 2019. At this time GOV.UK signed up for [Extended Security Maintenance](https://ubuntu.com/esm) (ESM) which provides continued security fixes for high and critical common vulnerabilities and exposures (CVEs) for the [most commonly used packages](https://wiki.ubuntu.com/SecurityTeam/ESM/14.04#A14.04_Infrastructure_ESM_Packages). ESM will end in April 2022.
+
+Running an old operating system is preventing GOV.UK from upgrading key components of our technology stack (MongoDB, Puppet, Python, ...).
+
+Upgrading Ubuntu in the existing infrastructure is difficult for a number of reasons, including but not limited to:
+
+- Multiple applications share the operating system of the same virtual machine. This means it’s not easy to upgrade a single application at a time. Furthermore, the way our [Puppet code](http://github.com/alphagov/govuk-puppet/) is written makes supporting virtual machines on different operating systems at the same time hard to support.
+- GOV.UK currently uses upstart as our init system, but newer versions of Ubuntu switch to systemd. This means GOV.UK will need to do a significant rewrite of our init scripts as part of an upgrade.
+- Our current version of Puppet is 3.8, which is no longer supported (the current version is 6.13). Support for newer operating system versions is likely to be lacking in unsupported versions of Puppet.
+- Other infrastructure as code tools in use (e.g. [Fabric](https://github.com/alphagov/fabric-scripts)) are also currently using very old versions which are very likely to have problems with newer operating systems.
+
+## Proposal
+
+GOV.UK will use a managed, container based hosting environment, such as [GOV.UK PaaS](https://www.cloud.service.gov.uk). The choice of hosting environment is out of scope for this RFC.
+
+GOV.UK applications and their dependencies will be built into container images. This may be done using buildpacks or other means - this decision is also out of scope for this RFC.
```

Comment:
I don't really understand "buildpacks", but I thought the buildpack approach was explicitly different from building images (or at least building images yourself). If that's the case, then this is a bit contradictory?

Reply:
is it specifically this -> https://buildpacks.io ?

Reply:
The terminology gets a bit confusing:

- currently in cloud foundry, buildpacks are used to build "droplets", which are a bit like an image that's missing the base layer
- droplets are combined with "stacks" (which are a bit like base layers) to form containers
- there are also [Cloud Native Buildpacks](https://buildpacks.io/), which _do_ build images (in OCI format)

In that light, buildpacks should really be seen as alternatives to Dockerfiles, not alternatives to docker images.

This sentence is really just here to make it clear that just because we're talking about containers, it doesn't necessarily mean we're excluding buildpacks from the discussion. It should probably say "containers" not "container images" though, to be fully correct.

```suggestion
GOV.UK applications and their dependencies will be built into containers. This may be done using buildpacks, Dockerfile or some other means - this decision is also out of scope for this RFC.
```

Reply:
> This sentence is really just here to make it clear that just because we're talking about containers, it doesn't necessarily mean we're excluding buildpacks from the discussion. It should probably say "containers" not "container images" though, to be fully correct.

At least currently, I'm not sure it's adding clarity. I assumed that part of this proposal was explicitly excluding some approaches to "containers" by excluding approaches that don't use images. If that is the case, it would be good to give at least one example of an excluded approach.

If I'm interpreting your view Richard, maybe saying something like "GOV.UK applications will deployed through containers to a managed, container based hosting environment" would be clearer and more concise?

Reply:
> I assumed that part of this proposal was explicitly excluding some approaches to "containers" by excluding approaches that don't use images. If that is the case, it would be good to give at least one example of an excluded approach.

That's not the intention, hence "or other means". The idea is just "we'll use a managed platform that runs containers for us". The RFC isn't trying to say anything about how we build the applications into containers - I think it's best to have a separate discussion about those (which probably wants to be informed by some technical research).

Reply:
I've tried to make this a little bit clearer as suggested above in e72f336.

Hopefully there will be a future RFC where we can discuss the buildpacks / cloud native buildpacks / Nix / Guix / Dockerfile decision. There's a lot to think about though, so let's keep it separate.